packages/lib/src/data.ts

```diff
@@ -0,0 +1,112 @@
+import { CID } from 'multiformats/cid'
+import { check, cidForCbor, HOUR } from '@atproto/common'
+import * as t from './types'
+import {
+  assureValidCreationOp,
+  assureValidOp,
+  assureValidSig,
+  normalizeOp,
+} from './operations'
+import {
+  ImproperOperationError,
+  LateRecoveryError,
+  MisorderedOperationError,
+} from './error'
+
+export const assureValidNextOp = async (
+  did: string,
+  ops: t.IndexedOperation[],
+  proposed: t.Operation,
+): Promise<{ nullified: CID[]; prev: CID | null }> => {
+  await assureValidOp(proposed)
+
+  // special case if account creation
+  if (ops.length === 0) {
+    await assureValidCreationOp(did, proposed)
+    return { nullified: [], prev: null }
+  }
+
+  const proposedPrev = proposed.prev ? CID.parse(proposed.prev) : undefined
+  if (!proposedPrev) {
+    throw new MisorderedOperationError()
+  }
+
+  const indexOfPrev = ops.findIndex((op) => proposedPrev.equals(op.cid))
+  if (indexOfPrev < 0) {
+    throw new MisorderedOperationError()
+  }
+
+  // if we are forking history, these are the ops still in the proposed canonical history
```

Comment:
I assume that we must support forking history.  But I wonder if it's also useful to be able to express _I don't want to fork history_, essentially just making `prev` a check that you've worked from the current version of the doc (a la compare-and-swap) rather than a race to fork history first.  Just a thought!